Add kafka support

welpaolo opened the pull request:
This draft PR includes the first steps to include Kafka in the data-integrator.

Some unit tests need to be fixed. 
The management of the config option could be improved.

deusebio commented on src/charm.py:
```diff
@@ -9,76 +9,181 @@
         super().__init__(*args)
 
-        self.database = self.model.config.get("database", "")
+        self.database = self.model.config.get("database-name", "")
```
This is a minor comment but I would allow database to have "None" value if it is not specified. 

I can see that it was implemented with "" before, so it is more of a suggestion up for you to follow up if you agree and if it is not too much of an hassle

deusebio commented on src/charm.py:
```diff
@@ -9,76 +9,181 @@
+            self.on["mongodb"].relation_broken, self._on_database_relation_broken
+        )
+
+        kafka_user_roles = KAFKA_EXTRA_USER_ROLES + self.extra_user_roles
```
Is it by design that the KAFKA_EXTRA_USER_ROLES cannot be overridden? WHy is it so?

I mean, could I get a user with "producer" permission only?

welpaolo replied:
I removed that

deusebio commented on src/charm.py:
```diff
@@ -90,12 +195,49 @@ def _on_get_credentials_action(self, event: ActionEvent) -> None:
+    def _on_database_relation_broken(self, _):
+        """Event triggered when the database relation is removed."""
+        if not self.is_database_related:
+            self.database = ""
```
Uhm, this does not seem right. I mean, I would not override the config we provide.

deusebio replied:
I mean, if we unrelate and then relate again, I would expect the same topic/database to be binded.

deusebio commented on src/charm.py:
```diff
@@ -90,12 +195,49 @@ def _on_get_credentials_action(self, event: ActionEvent) -> None:
+
+    def _on_kafka_relation_broken(self, _):
+        """Event triggered when kafka relatoion is removed."""
+        if not self.is_database_related:
```
```suggestion
        if not self.is_kafka_related:
```

deusebio commented on src/charm.py:
```diff
@@ -90,12 +195,49 @@ def _on_get_credentials_action(self, event: ActionEvent) -> None:
+    def _on_kafka_relation_broken(self, _):
+        """Event triggered when kafka relatoion is removed."""
+        if not self.is_database_related:
+            self.kafka = ""
```
```suggestion
            self.topic = ""
```

deusebio replied:
But really, I'm not sure we should set this to ""

deusebio replied:
I mean, if we unrelate and then relate again, I would expect the same topic/database to be binded.

deusebio commented on src/charm.py:
```diff
@@ -9,76 +9,181 @@
+            self.unit.status = BlockedStatus("The database name or topic name is not specified.")
+        elif self.database == "":
+            if self.is_kafka_related:
+                self.unit.status = ActiveStatus(f"The topic name is specified: {self.topic}.")
```
I believe the convention is to NOT add messages for ActiveStatus, if it is all good. But please @marcoppenheimer , double check this is correct.

deusebio commented on src/charm.py:
```diff
@@ -9,76 +9,181 @@
+        elif self.topic == "":
+            if self.is_database_related:
+                self.unit.status = ActiveStatus(
+                    f"The database name is specified: {self.database}."
```
I believe the convention is to NOT add messages for ActiveStatus, if it is all good. But please @marcoppenheimer , double check this is correct.

deusebio commented on src/charm.py:
```diff
@@ -9,76 +9,181 @@
+
+        elif event:
+            self.unit.status = ActiveStatus(
+                f"Both database name({self.database}) and topic name({self.topic}) has been specified!"
```
I believe the convention is to NOT add messages for ActiveStatus, if it is all good. But please @marcoppenheimer , double check this is correct.

deusebio commented on src/charm.py:
```diff
@@ -9,76 +9,181 @@
+            self.topic == new_topic
+        elif self.topic != new_topic:
+            self.unit.status = ActiveStatus(
+                f"New database name specified: {new_database_name}. Please remove current relation/s and add a new relation!"
```
I'm not sure the UX is right here. I mean: I would probably go into either Maintenance or Block status, since the application is really not functioning. Probably Block is more correct, since we are expecting the user to do something.  

https://juju.is/docs/olm/status-values

welpaolo replied:
I substitute it with blocked status

deusebio reviewed (changes requested):
Uhm, I have few comments I'd like to figure out. @marcoppenheimer could possibly help but if we still have doubts, we can then refer to Mykola

deusebio commented on src/charm.py:
```diff
@@ -90,12 +178,100 @@ def _on_get_credentials_action(self, event: ActionEvent) -> None:
+
+        return relation.data[self.unit]
+
+    def get_secret(self, scope: str, key: str) -> Optional[str]:
```
We should really factor these methods out to a common class in a charm library or similar....

We are doing a lot of copy and pasted every time. 

But let's address this item properly. No need to follow up on this within this PR

deusebio replied:
I don't understand your point. I mean, we could just have a MixinClass or a CharmBase-derived class in our charm or python libraries, with useful methods we generally use:

```python 

class DataPlatformCharm(CharmBase): #or mixin if we don't like inheritance

    def get_secret(...):
        # Implementation

    def set_secret(...):
        # Implementation

```

Why are we bounded to the OperatorFramework? I really don't see this going there or being rather hard to push these there. If - as dataplatform - we really like these methods, I'd factor them out in a lib (being charm or normal python). This sounds a lot better than copy and paste

marcoppenheimer replied:
These `set-secret` `get-secrets` are already meant to be removed as soon as an equivalent method exists in Operator Framework, and we added them only because it would make refactoring easier when that time came around if we had only 2 method names to reference. I would avoid investing more time on a stop-gap.

deusebio replied:
As said, no need to address this issue here. However, generally speaking, I can see value on having a process for implementing/experimenting things in data platform charm libraries (or packages) to be used within our projects first to then promote such features to the Operator Framework.

deusebio commented on src/charm.py:
```diff
@@ -47,38 +54,119 @@ def __init__(self, *args):
+        if self.database == "" or not self.is_database_related:
+            self.set_secret("app", "database", new_database_name)
+        elif self.database != new_database_name:
+            self.unit.status = WaitingStatus(
```
Sorry for being pedantic, but I really don't think it is a WaitingStatus. According to docs:

waiting =>  The charm is alive and well, and is waiting for some event to occur. No human intervention required.

I believe it makes more sense the blocked here:



blocked  =>  The charm is stuck. Human attention/intervention is required.

marcoppenheimer reviewed (changes requested):
I left a bunch of change suggestions. Overall I would like to see a reduction in the branching `if`s logic, which would become challenging to unpick/debug. We also aren't explicit with what we're checking often, especially with equivalency-checks to empty strings, which isn't intuitive. Personally I would use `if not X` and `.get(X, None)` more often, especially as setting empty-string values to relation/secret data will remove it from that dict/store.

I think more work is needed on abstracting a lot of these checks away elsewhere. 2 examples I saw often were checking for the existence of certain relations, or checking for changes in relation data field values. 

For the former, I think we should add some more checks/properties. Possibly having a `Database` object and a `Kafka` object with these checks built in? API could look something like `if self.Kafka and not self.Kafka.topic_changed`. Hopefully you get the idea.

For the latter, or generally, can be leaning more heavily on the `Diff` object from data-interfaces. Maybe passing that diff through the event?

deusebio commented on src/charm.py:
```diff
@@ -47,38 +54,119 @@ def __init__(self, *args):
-            self.unit.status = ActiveStatus(f"database: {self.database}")
+        """Handle on config changed event."""
+        # Read new parameters
+        new_database_name = self.model.config.get("database-name", "")
+        new_topic = self.model.config.get("topic-name", "")
+        new_extra_user_roles = self.model.config.get("extra-user-roles", "")
+
+        # check the
+        if self.database == "" or not self.is_database_related:
```
Uhm. I honestly don't like the `if self.database` or `if not self.database`. And relying on implicit behavior. I had (not only me) bad experiences with this in the past.

marcoppenheimer replied:
I agree to a point, but there are 2 main reasons not to. 

1. If you accidentally set `self.relation.data[self.app].update({"key": ""})` because we default to empty-strings, it removes the key from the data, which is undesired. 
2. It goes against PEP8 recommendations - https://peps.python.org/pep-0008/#programming-recommendations, specifically:

> For sequences, (strings, lists, tuples), use the fact that empty sequences are false:

and

> Don’t compare boolean values to True or False using ==:

deusebio replied:
Uhm, interesting. I did not know about this PEP nudge. Well, who the hell am I to disagree with PEP and Guido?! Yes, sure. Let's keep things in line with PEP. Thanks for posting this. Learnt something new.

welpaolo commented:
I made substantial changes to the PR and addressed the comments you made. 
I kept the unit message in case of the ActiveStatus, I think it is nice to show the configured parameters to the user. Let me know if there is any guideline against it.

deusebio commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+        if not self.unit.is_leader:
+            return
+        # update peer databag to trigger the charm status update
+        self._update_active_relations(event, "broken")
```
SUGGESTION

Instead of using strings ("broken", "active", ...), it would be good to use constants or enums. Does the ops framework provide something for this? @marcoppenheimer

deusebio replied:
sort of. I mean, if we use Literal, that helps for the typing only. But what I also meant was to also avoid the use of strings, rather replaced by constants.

deusebio commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+        # update parameter in the relation databag
+        self.set_secret("app", "database", new_database_name)
+        self.set_secret("app", "topic", new_topic)
+        self.set_secret("app", "extra-user-roles", new_extra_user_roles)
```
SUGGESTION

Best to use enum here, just to avoid having strings all over the place

deusebio replied:
As said before, this can help typing but not constant centralization

deusebio commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+        self.postgresql.database = self.database or ""
+        self.postgresql.extra_user_roles = self.extra_user_roles or ""
+        self.mongodb.database = self.database or ""
+        self.mongodb.extra_user_roles = self.extra_user_roles or ""
```
If no relations are defined all this code should not really be needed (since down in the update_database_relation we have an if statement that DOES things only if the relation exists). I believe it would REALLY improve readability to have "if" statement, making clear that you do these only if a relation exists

Some thing for Kafka later on

deusebio commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+                return WaitingStatus("Please provide database or topic name!")
+        # check configured database and emit status if needed
+        database_status = self._get_status_database(new_database_name)
+        if database_status:
```
SUGGESTION 

What about walrus operator

```
if (database_status := self._get_status_database(new_database_name)):
    return database_status
```

deusebio reviewed (approved):
Good to go! 

Just mentioning, that I found the logic in `_on_config_changed` a bit convoluted, but the UX we get is good and I personally don't have another way that would not imply a large restructuring which is clearly outside of the scope of this PR. 

I might think if there is a more linear (and functoinal) way to do this. I would eventually first design it, propose it and see whether it would improve the readability of the code or not

Some SUGGESTION attached

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+        if database_name:
+            if database_name != self.database:
+                if self.is_database_related:
+                    return BlockedStatus(
```
**todo:** Also type-hint return-type here.
**nitpick:** I'm not certain it's clear what this method does from the name+description alone. I would imagine it would return more than just `Blocked` or `None`. Maybe rename to something like `_has_database_changed()`? If I understood it's intent correctly. If so, maybe return a `bool` and have whatever calls it set the status, rather than it being indirectly generated here?



And for @deusebio's comment, maybe something like:

```python
def _get_status_database(self, database_name: str) -> Optional[StatusBase]:
    """Return unit status based on the database parameter read from config."""
    if database name and database_name != self.database and self.is_database_related:
        return BlockedStatus()

    return None

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+                    )
+        return None
+
+    def _get_status_topic(self, topic_name: str):
```
**todo:** Return type-hints here please.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+        """Return unit status based on the topic parameter read from config."""
+        if topic_name:
+            if topic_name != self.topic:
+                if self.is_kafka_related:
```
Same as previous.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+                    )
+        return None
+
+    def _get_status_extra_user_roles(self, extra_user_roles: str):
```
**todo:** Return type-hints here please.

marcoppenheimer commented on src/charm.py:
```diff
@@ -47,38 +54,119 @@ def __init__(self, *args):
+        # output the correct message.
+        self.set_status(event)
+
+    def update_database_relations(self, database_relation_data: Dict):
```
**todo:** Return type-hints here please. And please type the `Dict`

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
         if mysql:
-            result["mysql"] = list(mysql.values())[0]
+            result[MYSQL] = list(mysql.values())[0]
         if postgresql:
-            result["postgresql"] = list(postgresql.values())[0]
+            result[POSTGRESQL] = list(postgresql.values())[0]
         if mongodb:
-            result["mongodb"] = list(mongodb.values())[0]
+            result[MONGODB] = list(mongodb.values())[0]
+        if kafka:
+            result[KAFKA] = list(kafka.values())[0]
```
**chore:** Another candidate for a loop over some `CHARMS` data structure.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+            return
+        removed_relations = []
+        # check for relation that has been removed
+        for k, v in self.app_peer_data.items():
+            if "relation-" in k:
+                if v == "broken":
+                    removed_relations.append(k)
```
**chore:** Rename `k/v` to something descriptive. e.g `relation_data_key`, `value`

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+            for relation in removed_relations:
+                self.set_secret("app", relation, "removed")
+
+    @property
+    def database(self):
+        """Return the configured database name."""
+        return self.get_secret("app", "database")
+
+    @property
+    def topic(self):
+        """Return the configured topic name."""
+        return self.get_secret("app", "topic")
+
+    @property
+    def extra_user_roles(self):
+        """Return the configured user-extra-roles parameter."""
+        return self.get_secret("app", "extra-user-roles")
+
+    @property
+    def is_database_related(self):
+        """Return if a relation with database is present."""
+        possible_relations = [
+            self._check_for_credentials(self.mysql.relations),
+            self._check_for_credentials(self.postgresql.relations),
+            self._check_for_credentials(self.mongodb.relations),
+        ]
+        return any(possible_relations)
```
**todo:** Return type-hint please.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+        """Return if a relation with kafka is present."""
+        return self._check_for_credentials(self.kafka.relations)
+
+    @property
+    def app_peer_data(self) -> Dict:
+        """Application peer relation data object."""
+        relation = self.model.get_relation(PEER)
+        if not relation:
+            return {}
+
+        return relation.data[self.app]
+
+    @property
+    def unit_peer_data(self) -> Dict:
+        """Peer relation data object."""
+        relation = self.model.get_relation(PEER)
+        if relation is None:
+            return {}
+
+        return relation.data[self.unit]
```
**nitpick:** These aren't strictly-speaking `Dict` types. Change to `MutableMapping[str, str]`

marcoppenheimer reviewed (changes requested):
I think before it's possible to give this a high-level review, I think there needs to be some encapsulation and tidying first. Almost all of my comments are essentially three things:

1. Stricter type-hinting, both on arguments and return types
2. Avoid nested if-statements where possible
3. Strongly recommend building an easily extendable, Iterable data structure that maps the following:
    - Name of the product - e.g `postgresql`
    - Requires object - e.g `DatabaseRequires`
    - Relation-data for requires object args - e.g `database`, `extra-user-roles`, `topic` 

For point 3, look for opportunities to implicitly loop over this data structure wherever possible, and using the `Diff` provided by `data_interfaces` in events

e.g You're wanting the charm to update relation data on `config_changed` event:

```python
def _on_config_changed(event: ConfigChangedEvent) -> None:
    for charm, requires in MAPPING.items():
        requires._update_relation_data()
```

deusebio commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+        # update peer databag to trigger the charm status update
+        self._update_active_relations(event, "broken")
+
+    def _set_config_parameters(self, status: StatusBase):
+        """Return if the new parameters must be set."""
+        if isinstance(status, ActiveStatus):
+            return True
+        if isinstance(status, BlockedStatus) and "add relation" in status.message:
+            return True
+        return False
```
yes, I agree with @marcoppenheimer . It does not seem good to implement logic based on the juju status which is more of an output of our charm code, but not an input. Beside it that could also be changed by code other than ours, that can indeed create problems

deusebio commented on src/charm.py:
```diff
@@ -9,93 +9,377 @@
+            return True
+        return False
+
+    def _on_config_changed(self, _):
```
I don't mind typing the arg. I mean "_" is generally used as placeholder that is not used later.

welpaolo commented on metadata.yaml:
```diff
@@ -2,13 +2,18 @@
+  data-integrator-peers:
+    interface: data-integrator-peers
+
 requires:
```
yes this is fine

welpaolo commented on metadata.yaml:
```diff
@@ -2,13 +2,18 @@
 summary: |
-  Integrator charm for getting database credentials for non-native applications.
-  Not all applications implement a database relation interface but allow setting
+  Integrator charm for getting credentials for non-native applications.
```
Only getting

deusebio reviewed (commented):
I think this PR is shaping up very well. I only have one comment about the strutcute of the databag:

I believe it would make a very nice and neat databag to have it done not by relation-id, but by relation type, like:

```
MYSQL: active/broken/missing
KAFKA: active/broken/missing
MONGODB: active/broken/missing
POSTGRESQL: active/broken/missing
```

This would also give us a reliable and consistent view on which relation is active (and it could also prevent us from building more than 1 relation per-type)

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+            return BlockedStatus("Please relate the data-integrator with the desired product")
+
+        if self.is_kafka_related and self.topic_active != self.topic_name:
+            logger.error("Trying to change Kafka configuration for existing relation")
```
**todo:** The log message could include the message in the `BlockedStatus` below.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+            logger.error("Trying to change Kafka configuration for existing relation")
+            return BlockedStatus("To change topic, please remove relation and add it again")
+
+        if self.is_database_related and any(
+            [database != self.database_name for database in self.databases_active.values()]
+        ):
+            logger.error("Trying to change database-name configuration for existing relation")
+            return BlockedStatus(
+                "To change database name, please remove relation and add it again"
+            )
```
**suggestion:** Maybe it would be useful to loop over `self.databases_active.values()` at the top level, so that you can can extract `database-name`, and add that to the logging strings.

deusebio replied:
+1 using walrus if the code does not become unreadable. @welpaolo trust your judgement. If it is too unreadable in your opinion, don't use walrus

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+            }
+            self._update_database_relations(database_relation_data)
 
-        for rel in self.mysql.relations:
-            self.mysql._update_relation_data(rel.id, {"database": self.database})
-        for rel in self.postgresql.relations:
-            self.postgresql._update_relation_data(rel.id, {"database": self.database})
-        for rel in self.mongodb.relations:
-            self.mongodb._update_relation_data(rel.id, {"database": self.database})
+        if not self.topic_active and self.topic_name:
+            for rel in self.kafka.relations:
+                self.kafka._update_relation_data(
+                    rel.id,
+                    {
+                        "topic": self.topic_name or "",
+                        "extra-user-roles": self.extra_user_roles or "",
+                    },
```
**thought:** Non-blocking, just an idea, I wonder if this could take a `DefaultDict`, do some clever star-expansion from `**self`, and any missing/None attrs defaults dict values to to `""`. Complicated and not needed though.

deusebio replied:
I'd defer this later. This PR is already quite big to also put extra features

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+    def _update_database_relations(self, database_relation_data: Dict[str, str]):
+        """Update the relation data of the related databases."""
+        for db_name, relation in self.database_relations.items():
+            logger.debug(f"Update databag for database: {db_name}")
```
**todo:** Would use the present-tense `Updating databag...` here.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+        removed_relations = []
+        # check for relation that has been removed
+        for relation_data_key, relation_value in self.app_peer_data.items():
+            if "relation-" in relation_data_key:
+                if relation_value == "broken":
```
**todo:** Would use `if "relation-" in relation_data_key and relation_value == "broken"` here, as I don't *think* they need to be nested.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+
+        return ActiveStatus()
+
+    def _on_config_changed(self, _) -> None:
```
**todo:** I would type-hint the arg here. Even if it's unused, we're calling it on L186 so it's helpful to know that it's `EventBase` and takes any event.

deusebio replied:
I don't mind. But I'd rather avoid to call _on_config_changed in L186

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+    @property
+    def kafka_relation(self) -> Optional[Relation]:
+        """Return the kafka relation if present."""
+        return self.kafka.relations[0] if len(self.kafka.relations) > 0 else None
```
**nitpick:** This can be `return self.kafka.relations[0] if len(self.kafka.relations) else None`. 0 is falsie.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+        return {
+            name: requirer.relations[0]
+            for name, requirer in self.databases.items()
+            if len(requirer.relations) > 0
```
**nitpick:** Same as above.

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+        return any(possible_relations)
+
+    @staticmethod
+    def _check_for_credentials(relations: List[Relation]) -> bool:
+        """Check if credentials are present in the relation databag."""
+        for relation in relations:
+            if (
+                "username" in relation.data[relation.app]
+                and "password" in relation.data[relation.app]
+            ):
+                return True
+        return False
```
**question:** I'm sure it's fine, but just checking, do we mean to exit on first relation found with username/password? 

What about the other relations? Or is it assumed that any list we pass this function, if one is set, they all are?

deusebio replied:
I believe we have only 1 relation per type. See the metadata.yaml

https://github.com/canonical/data-integrator/blob/main/metadata.yaml

But I'm not sure what Juju does if we try to relate multiple times on the same interface. Would it complain?

welpaolo replied:
Yes It will complain, Juju will allow only one relation per interface.

marcoppenheimer commented on tests/unit/test_charm.py:
```diff
@@ -14,25 +14,33 @@ def setUp(self):
 
         self.assertEqual(
             action_event.fail.call_args,
-            [("The database name is not specified in the config.",)],
+            [("The database name or topic name is not specified in the config.",)],
         )
```
**issue:** I would not test based on non-`constants`/`literals` strings, as these could change any time, slowing down refactoring for failing tests. Maybe here check for the *existence* of `action_event.fail` and `action_event.fail.call_args`, making sure there is *some* message, rather than exactly that one.

deusebio replied:
If possible, I'd go with Marc suggestion. Otherwise we can just keep it as is for the time being.

marcoppenheimer commented on tests/unit/test_charm.py:
```diff
@@ -42,36 +50,121 @@ def test_action_failures(self):
+        self.harness.update_config({"extra-user-roles": "admin"})
         self.harness.charm._on_config_changed(Mock())
         self.assertEqual(
-            self.harness.model.unit.status, BlockedStatus("The database name is not specified.")
+            self.harness.model.unit.status,
+            BlockedStatus("Please relate the data-integrator with the desired product"),
```
**issue:** Same as above.

marcoppenheimer commented on tests/unit/test_charm.py:
```diff
@@ -42,36 +50,121 @@ def test_action_failures(self):
+            },
+        )
+        self.harness.charm._on_config_changed(Mock())
+        self.assertEqual(
+            self.harness.model.unit.status,
+            ActiveStatus(""),
+        )
```
**issue:** For `ActiveStatus` we should not have a message, so here I would remove the `""`, but the `assertEquals` is good here.

marcoppenheimer commented on tests/unit/test_charm.py:
```diff
@@ -42,36 +50,121 @@ def test_action_failures(self):
+        )
+        self.harness.update_config({"topic-name": "bar1"})
+        self.harness.charm._on_config_changed(Mock())
+        self.assertEqual(
+            self.harness.model.unit.status,
+            BlockedStatus("To change topic, please remove relation and add it again"),
```
**issue:** Same as other comments.

marcoppenheimer commented:
@welpaolo - Forgot to comment the review. Overall I think this is MUCH tidier and I really like it! I left a bunch of very small points, but once the **todo**s and **issue**s are addressed I'm happy to approve straight away. **question**s and **thought**s are non-blocking, just if we want a discussion, but I'm aware this has gone on for a while so happy to just approve and merge :)

deusebio commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
-        # MongoDB
-        self.mongodb = DatabaseRequires(self, relation_name="mongodb", database_name=self.database)
-        self.framework.observe(self.mongodb.on.database_created, self._on_database_created)
+    def get_status(self) -> StatusBase:
+        """Return the current application status."""
```
I would rather call the method within the update_status, i.e.:

```
def update_status(_):
    self.unit.status = self.get_status()
```

Note that this method only computes the status, does not set it. I would leave the unpure logic with side effects (setting the unit status), clear in the endpoint implementation, while factoring out the business logic (input => output) for figuring out the status

marcoppenheimer commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+
+        if removed_relations:
+            # update configuration
+            self._on_config_changed(event)
```
Isn't the intent to also update the relations, hence why we're calling `on_config_changed` here?

I'm for adding after a `get_status()` though.

deusebio commented on tests/unit/test_charm.py:
```diff
@@ -42,36 +50,121 @@ def test_action_failures(self):
+        self.harness.set_leader(True)
+        self.harness.update_config({"database-name": "foo"})
+        self.harness.charm._on_config_changed(Mock())
+        self.assertEqual(
+            self.harness.model.unit.status,
+            BlockedStatus("Please relate the data-integrator with the desired product"),
```
I suppose this would miss the point of testing that the "right" blocked message is created. We have different block status (depending on the reason). We want to tests we get the one we want I suppose

welpaolo replied:
I will follow the suggestion done by @deusebio

welpaolo commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+        # check for relation that has been removed
+        for relation_data_key, relation_value in self.app_peer_data.items():
+            if "relation-" in relation_data_key:
+                if relation_value == "broken":
```
Done

welpaolo commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
-        if self.model.config.get("database", "") == "":
-            event.fail("The database name is not specified in the config.")
+        if not self.database_name and not self.topic_name:
+            event.fail("The database name or topic name is not specified in the config.")
             event.set_results({"ok": False})
```
Yes the status is set.

welpaolo commented on src/charm.py:
```diff
@@ -9,93 +9,306 @@
+        return any(possible_relations)
+
+    @staticmethod
+    def _check_for_credentials(relations: List[Relation]) -> bool:
```
I will leave it like it is for the moment.

deusebio reviewed (approved):
Very good work! Kudos @welpaolo
